components/List/src/MDCSelfSizingStereoCell.m

```diff
@@ -0,0 +1,464 @@
+
+- (void)createSubviews {
+  self.textContainer = [[UIView alloc] init];
+  [self.contentView addSubview:self.textContainer];
+
+  self.titleLabel = [[UILabel alloc] init];
+  self.titleLabel.font = self.defaultTitleLabelFont;
+  [self.textContainer addSubview:self.titleLabel];
+
+  self.detailLabel = [[UILabel alloc] init];
+  self.detailLabel.font = self.defaultDetailLabelFont;
+  [self.textContainer addSubview:self.detailLabel];
+
+  self.leadingImageView = [[UIImageView alloc] init];
+  [self.contentView addSubview:self.leadingImageView];
+
+  self.trailingImageView = [[UIImageView alloc] init];
+  [self.contentView addSubview:self.trailingImageView];
+}
+
+#pragma mark UIView Overrides
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  MDCSelfSizingStereoCellLayout *layout = [self layoutForCellWidth:self.frame.size.width];
+  self.textContainer.frame = layout.textContainerFrame;
+  self.titleLabel.frame = layout.titleLabelFrame;
+  self.detailLabel.frame = layout.detailLabelFrame;
+  self.leadingImageView.frame = layout.leadingImageViewFrame;
+  self.trailingImageView.frame = layout.trailingImageViewFrame;
+  if (self.mdf_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    self.titleLabel.textAlignment = NSTextAlignmentRight;
+    self.detailLabel.textAlignment = NSTextAlignmentRight;
+    self.leadingImageView.frame =
+        MDFRectFlippedHorizontally(self.leadingImageView.frame, layout.cellWidth);
+    self.trailingImageView.frame =
+        MDFRectFlippedHorizontally(self.trailingImageView.frame, layout.cellWidth);
+    self.textContainer.frame =
+        MDFRectFlippedHorizontally(self.textContainer.frame, layout.cellWidth);
+    self.titleLabel.frame =
+        MDFRectFlippedHorizontally(self.titleLabel.frame, self.textContainer.frame.size.width);
+    self.detailLabel.frame =
+        MDFRectFlippedHorizontally(self.detailLabel.frame, self.textContainer.frame.size.width);
+  } else {
+    self.titleLabel.textAlignment = NSTextAlignmentLeft;
+    self.detailLabel.textAlignment = NSTextAlignmentLeft;
+  }
+}
+
+- (void)setNeedsLayout {
+  [self invalidateCachedLayouts];
+  [super setNeedsLayout];
+}
+
+- (CGSize)systemLayoutSizeFittingSize:(CGSize)targetSize {
+  MDCSelfSizingStereoCellLayout *layout = [self layoutForCellWidth:targetSize.width];
+  return CGSizeMake(targetSize.width, layout.calculatedHeight);
+}
+
+#pragma mark UICollectionViewCell Overrides
+
+- (void)prepareForReuse {
+  [super prepareForReuse];
+
+  [self invalidateCachedLayouts];
+  self.titleLabel.text = nil;
+  self.titleLabel.font = self.defaultTitleLabelFont;
+  self.detailLabel.text = nil;
+  self.detailLabel.font = self.defaultDetailLabelFont;
+  self.leadingImageView.image = nil;
+  self.trailingImageView.image = nil;
+}
+
+#pragma mark Layout
+
+- (MDCSelfSizingStereoCellLayout *)layoutForCellWidth:(CGFloat)cellWidth {
+  MDCSelfSizingStereoCellLayout *layout = self.cachedLayouts[@(cellWidth)];
```

Comment:
I'm curious whether using a CGFloat as a key may prove finicky. Should we cast to an int?

Reply:
I don't _think_ it'd be an issue, but it's probably a good idea to floor it just in case 👍

Reply:
Done in 64652f0f58c0521445e92d50910e32fb9bdee360